refactor(signing): use keymanager trait instead of exposing secret ke…

oetyng opened the pull request:
…y share
Required by downstream sn consumer.

oetyng committed:
refactor(signing): use keymanager trait instead of exposing secret key share

dan-da commented on examples/mint-repl/mint-repl.rs:
```diff
@@ -185,17 +188,19 @@ fn mk_new_random_mint(threshold: usize, amount: u64) -> Result<MintInfo> {
-        let key_manager = KeyManager::new(
-            secret_key_set.public_keys().clone(),
-            (i, secret_key_set.secret_key_share(i).clone()),
+        let key_manager = Arc::new(KeyManager::new(
```
why does this need to use Arc?

oetyng replied:
Due to line 189 in `mint.rs`:

```rust
    pub fn verifier(&self) -> Verifier<K> {
        Verifier::new(self.key_manager.clone())
    }
```

dan-da replied:
ok, I see there is:

```
pub struct Verifier<K: KeyManager> {
    key_manager: Arc<K>,
}

impl<K: KeyManager> Verifier<K> {
    pub fn new(key_manager: Arc<K>) -> Self {
        Self { key_manager }
    }
```

I just don't understand yet the reason for (a) using Arc in sn_dbc, and (b) exposing/requiring it at API level, ie Verifier::new(km: Arc<K>)

I guess I'm just wondering if we are mixing concerns here.  Maybe this is anticipating changes in the async PR?

oetyng replied:
It's to be able to return a `Verifyer` from the `Mint`., as to pass in something with a limited API to the dbc when confirming valid.
That was the original design, and in the discussion it was said to be better not to pass in `KeyManager` there.

Now that we don't actually hold secret keys, we only hold a reference to something that updates under the hood.
Unless we want to use that thing directly this is sort of the thing I'd say, as we can't have a verifier with a reference to keymanager without it. At least this is a simple way.

But I can just pass in the `KeyManager` straight away, if exposing that fulll api within DBC is less of a problem than Arc.

oetyng replied:
Pushed a commit with that approach now.

oetyng committed:
refactor(keymanager): return result on all the trait methods

oetyng committed:
fix(tests): also match on the mapped errors
The tests are setup to verify that a specific error is returned for a
specific failure condition. With the associated type in the trait, we
do not know the Error implementation type, and therefore map it
into a catch-all "Signing" variant. To not break the tests, we extract
the expected errors out of the enum variant when matching.

oetyng committed:
refactor: remove verifier struct

davidrusu commented on src/dbc.rs:
```diff
@@ -30,13 +31,15 @@ impl Dbc {
     // Check there exists a DbcTransaction with the output containing this Dbc
     // Check there DOES NOT exist a DbcTransaction with this Dbc as parent (already minted)
-    pub fn confirm_valid(&self, key_cache: &KeyCache) -> Result<(), Error> {
+    pub fn confirm_valid<K: KeyManager>(&self, verifier: &K) -> Result<(), Error> {
```
Another approach would be to have `V: Verifier` here and split up the `KeyManager` trait into two traits:

```rust
trait DbcSigner {
    ... signing functions
}
trait DbcVerifier {
    fn verify(..) {}
}
impl DbcSigner for SimpleKeyManager {
    ...
}
impl  DbcVerifier for SimpleKeyManager {
    ...
}
```
That way you can expose only the verifier api to the Dbc::confirm_valid despite using the same struct behind the scenes for signing as well.

davidrusu commented on src/key_manager.rs:
```diff
@@ -13,99 +13,133 @@ use threshold_crypto::{serde_impl::SerdeSecret, SecretKeyShare, SignatureShare};
 
 #[derive(Debug, Clone, Hash, PartialEq, Eq, Deserialize, Serialize)]
-pub struct NodeSignature(u64, SignatureShare);
+pub struct NodeSignature {
+    index: u64,
+    sig: SignatureShare,
+}
```
nice :)

davidrusu reviewed (approved):
Looks good to me :ship:
